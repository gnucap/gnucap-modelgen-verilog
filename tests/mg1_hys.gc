
attach ../modelgen_0.so

verilog

`modelgen
module hys(p, n);
    inout p, n;
    electrical p, n, ns;

    parameter real R = 1e3 from (0:inf);
    parameter real k = 1 from (0:inf);
    parameter real tau = 1e-5 from (0:inf);

    (* desc="state" *) real s;

    analog begin
        s = V(ns, n);
        I(p, n)  <+ V(p, n)/R * (1+tanh(k*s));
        I(ns, n) <+ V(p, n) - pow(s, 3) + s;
        I(ns, n) <+ - ddt(tau*s);
    end

endmodule

!make hys.so >/dev/null
attach ./hys.so

hys #() h1(1 0)

spice
.options noinsensitive
V1 1 0 1 sin(0 0.7 1k)

.options noinsensitive

.print tran v(1) i(V1) s(h1)
.tran 2m
* > hys.out

.status notime
