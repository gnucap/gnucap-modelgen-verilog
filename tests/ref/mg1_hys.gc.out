#Time       v(1)       i(V1)      s(h1)     
 0.         0.         0.         0.        
 11.854n    52.137u   -52.137n    61.876n   
 23.708n    104.27u   -104.27n    154.71n   
 43.234n    190.15u   -190.15n    442.74n   
 62.76n     276.03u   -276.03n    899.19n   
 189.95n    835.45u   -835.46n    8.0245u   
 415.36n    0.0018268 -1.8269u    38.555u   
 833.07n    0.003664  -3.6646u    157.33u   
 1.4798u    0.0065086 -6.5119u    507.8u    
 2.1266u    0.0093531 -9.3631u    0.0010718 
 3.3714u    0.014827  -14.869u    0.0028189 
 4.6162u    0.0203    -20.412u    0.0055245 
 6.6606u    0.029286  -29.65u     0.012428  
 8.7051u    0.038268  -39.146u    0.022949  
 10.75u     0.047243  -49.033u    0.037904  
 13.963u    0.061335  -65.811u    0.073118  
 17.177u    0.075401  -84.913u    0.12683   
 20.39u     0.089437  -107.51u    0.20491   
 23.604u    0.10344   -134.77u    0.31278   
 29.034u    0.12699   -191.14u    0.55623   
 37.755u    0.1645    -283.56u    0.91548   
 42.974u    0.18672   -331.2u     1.0296    
 48.192u    0.20874   -373.82u    1.0738    
 62.336u    0.26721   -482.13u    1.1107    
 69.62u     0.29653   -535.89u    1.1189    
 81.761u    0.34399   -623.7u     1.1361    
 100.21u    0.41218   -750.57u    1.1598    
 131.35u    0.51432   -941.99u    1.193     
 241.27u    0.69895   -0.0012916  1.2487    
 351.19u    0.56322   -0.0010349  1.2126    
 557.29u   -0.24658    418.67u    0.86322   
 590.93u   -0.37853    617.21u    0.74235   
 613.2u    -0.45695    705.48u    0.60966   
 627.54u   -0.5028     716.94u    0.45488   
 634.64u   -0.52402    686.44u    0.32048   
 641.74u   -0.5442     580.57u    0.066927  
 647.15u   -0.55886    404.18u   -0.28418   
 651.23u   -0.56946    246.2u    -0.64408   
 655.27u   -0.57962    146.73u   -0.96582   
 659.32u   -0.58941    108.6u    -1.144     
 663.36u   -0.59882    98.949u   -1.2036    
 669.79u   -0.61296    97.605u   -1.2238    
 674.41u   -0.62251    98.719u   -1.226     
 679.02u   -0.63154    99.672u   -1.2286    
 689.26u   -0.64965    101.53u   -1.2339    
 719.75u   -0.68739    105.25u   -1.245     
 811.19u   -0.64889    100.97u   -1.2363    
 981.02u   -0.083269   18.26u    -1.0472    
 0.0010659  0.28178   -88.93u    -0.83735   
 0.0010993  0.40894   -162.07u   -0.69893   
 0.0011166  0.46823   -222.76u   -0.58217   
 0.0011279  0.50396   -289.7u    -0.45397   
 0.0011392  0.53715   -454.69u   -0.15474   
 0.0011471  0.55876   -739.77u    0.33606   
 0.001152   0.57155   -940.02u    0.76616   
 0.0011567  0.5831    -0.0010437  1.0712    
 0.0011613  0.59416   -0.001088   1.1917    
 0.001166   0.60471   -0.001112   1.2175    
 0.0011706  0.61474   -0.0011315  1.2233    
 0.0011753  0.62425   -0.0011496  1.2264    
 0.0011837  0.64012   -0.0011797  1.2311    
 0.0012014  0.66766   -0.001232   1.2392    
 0.0012484  0.69997   -0.0012935  1.2491    
 0.0013987  0.41596   -758.63u    1.1685    
 0.001549  -0.21232    363.27u    0.88917   
 0.0016086 -0.44134    687.37u    0.62914   
 0.0016224 -0.48674    716.11u    0.51166   
 0.0016308 -0.5128     702.32u    0.38795   
 0.0016361 -0.52824    665.53u    0.26599   
 0.0016413 -0.54311    579.47u    0.067041  
 0.0016466 -0.55739    410.29u   -0.27031   
 0.0016504 -0.56724    262.21u   -0.60098   
 0.0016543 -0.57725    155.9u    -0.92858   
 0.0016583 -0.58692    111.92u   -1.125     
 0.0016622 -0.59622    99.676u   -1.1973    
 0.0016716 -0.61679    97.304u   -1.2288    
 0.0016763 -0.62628    99.045u   -1.2275    
 0.0016833 -0.63949    100.5u    -1.2309    
 0.0016904 -0.65145    101.71u   -1.2345    
 0.0017067 -0.67422    103.98u   -1.2411    
 0.0017433 -0.69939    106.34u   -1.2488    
 0.0018717 -0.50523    84.689u   -1.1958    
 0.002      0.        -13.499p   -1.0099    
Gnucap   System status
iterations: op=0, dc=0, tran=735, fourier=0, total=833
transient timesteps: accepted=83, rejected=24, total=107
nodes: user=1, subckt=0, model=1, total=2
dctran density=100.0%, ac density=100.0%
